virtualizarr/readers/safetensors.py

```diff
@@ -0,0 +1,641 @@
+                if len(custom_names) != len(shape):
+                    raise ValueError(
+                        f"Provided dimension names for '{tensor_name}' has {len(custom_names)} "
+                        f"names, but tensor has {len(shape)} dimensions."
+                    )
+                dim_names = custom_names
+            else:
+                dim_names = [f"{tensor_name}_dim_{i}" for i in range(len(shape))]
+
+            tensor_attrs = {}
+            #  not clear to me from the spec if additional keys allowed, parse just in case
+            for key, value in tensor_info.items():
+                if key not in {"dtype", "shape", "data_offsets"}:
+                    tensor_attrs[key] = (
+                        value if isinstance(value, str) else json.dumps(value)
+                    )
+
+            tensor_attrs["original_safetensors_dtype"] = dtype_str
+
+            tensor_attrs["safetensors_storage_info"] = json.dumps(
+                {
+                    "chunked": False,
+                    "contiguous": True,
+                }
+            )
+
+            metadata = create_v3_array_metadata(
+                shape=shape,
+                data_type=dtype,
+                chunk_shape=shape,  # Treat the whole tensor as a single chunk
+                dimension_names=dim_names,
+                attributes=tensor_attrs,
+            )
+
+            manifest_array = ManifestArray(
+                metadata=metadata,
+                chunkmanifest=chunk_manifest,
+            )
+
+            manifest_dict[tensor_name] = manifest_array
+
+        return ManifestGroup(arrays=manifest_dict, attributes=attrs)
+
+    @staticmethod
+    def _create_manifest_store(
+        filepath: str,
+        drop_variables: list,
+        dimension_names: Optional[Dict[str, list[str]]] = None,
+        revision: Optional[str] = None,
+    ) -> ManifestStore:
+        """
+        Create a ManifestStore for a SafeTensors file.
+
+        This method handles the complete workflow of reading a SafeTensors file and
+        creating a virtual store. It automatically determines the appropriate store type
+        based on the filepath and handles Hugging Face Hub URLs specially.
+
+        Parameters
+        ----------
+        filepath : str
+            Path to the SafeTensors file. Can be:
+            - Local filesystem path (e.g., "/path/to/file.safetensors")
+            - HTTP/HTTPS URL (e.g., "https://huggingface.co/.../file.safetensors")
+        drop_variables : list
+            List of tensor names to exclude from the dataset.
+        dimension_names : Dict[str, list[str]], optional
+            Custom dimension names for specific tensors. Keys are tensor names,
+            values are lists of dimension names matching the tensor's shape.
+        revision : str, optional
+            Repository revision (branch, tag, or commit hash) for Hugging Face Hub.
+            Defaults to "main" if not specified.
```

Comment:
Interesting. This begs the question of whether we could virtualize not just the current snapshot of SafeTensors data in Icechunk, but the entire version history...

Reply:
Yes, HF Hub is really just a git repo. It has a lot of similarities with IceChunk. See for example https://huggingface.co/blog/xet-on-the-hub and all the linked previous articles at the bottom of the post.